<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <script src="shared/jquery.js" type="text/javascript"></script>
    <script src="shared/shiny.js" type="text/javascript"></script>
    <link rel="stylesheet" type="text/css" href="shared/shiny.css"/>
    <!--
    <link rel="stylesheet" type="text/css" href="shared/slider/css/jquery.slider.min.css"/>
    <script src="shared/slider/js/jquery.slider.min.js"></script>
    -->


    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap.min.css"/>
    <script src="shared/bootstrap/js/bootstrap.min.js"></script>
    <!--
    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap-responsive.min.css"/>
    -->

    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <script src="d3.v3.min.js" charset="utf-8"></script>
    <script src="gam.js" type="text/javascript"></script>
    <link rel="stylesheet" type="text/css" href="gam.css"/>
    <script> (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){ (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o), m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m) })(window,document,'script','//www.google-analytics.com/analytics.js','ga'); ga('create', 'UA-44771306-1', 'wustl.edu'); ga('send', 'pageview'); </script>
    <title>Shiny GAM</title>
  </head>
  <body>
  <div id="loading">
    <div id="loading-dialog">
        <span>
            <p>Please wait for operation to finish...</p>
            <img src="loader.gif"></img>
        </span>
    </div>
  </div>
      <div id="networkParameters" class="js-output">
          <script>
              network = {
                  available: false,
                  hasReactionsAsEdges: false,
                  hasReactionsAsNodes: false,
                  hasGenes: false,
                  usesRpairs: false
              };
          </script>
      </div>
      <div id="moduleParameters" class="js-output">
          <script>
              module = {
                  available: false
              };
          </script>
      </div>
    <div class="container">
      <div class="row">
        <div class="col-md-12" style="padding: 10px 0px;">
          <h1>Shiny GAM: integrated analysis of genes and metabolites</h1>
        </div>
      </div>
      <div class="row">
          <div class="col-md-12">
              <div class="tabbable">
                  <ul class="nav nav-tabs" id="tab-panel-main">
                      <!--<li class="active"> <a href="#tab-home" data-toggle="tab">Home</a> </li>-->
                      <li class="active"> <a href="#tab-work" data-toggle="tab">Work</a> </li>
                      <li> <a href="#tab-help" data-toggle="tab">Help</a> </li>
                      <li> <a href="#tab-about" data-toggle="tab">About</a> </li>
                  </ul>
                  <div class="tab-content">
                      <!--
                      <div class="tab-pane active" id="tab-home">
                      </div>
                      -->
                      <div class="tab-pane active" id="tab-work">
                          <div class="row">
                              <div class="col-md-12">
                                  <div class="alert alert-info" role="alert">If you have any feedback, please send it to <a href="mailto:asergushichev@path.wustl.edu">asergushichev@path.wustl.edu</a></div>
                              </div>
                          </div>
                          <div class="row">
                              <div class="col-md-3">
                                  <form class="well">
                                      <p class="form-header">Construct a network</p>
                                      <label class="control-label" for="network">
                                          Select an organism: 
                                          <a id="select-organism-tooltip" class="help-tooltip" href="#" 
                                              title="Select an organism that you want to analyse. If there is no one that you want contact the author (see About tab).">
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                      <select id="network" class="selectpicker">
                                          <option value="Mouse musculus" selected="selected">Mouse musculus</option>
                                          <option value="Homo sapiens">Homo sapiens</option>
                                      </select>
                                      <label>
                                          File with gene DE data: 
                                          <a id="upload-gene-data-tooltip" class="help-tooltip" href="#" 
                                              title='File with gene differential expression data shoud be tab-separated with header and contain columns "ID" and "pval".'>
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                      <input id="geneDE" type="file"/>
                                      <label>
                                          File with metabolic DE data: 
                                          <a id="upload-met-data-tooltip" class="help-tooltip" href="#" 
                                              title='File with metabolic differential expression data shoud be tab-separated with header and contain columns "ID" and "pval".'>
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                      <input id="metDE" type="file"/>
                                      <label class="control-label" for="reactionsAs">
                                          How to interpret reactions:
                                          <a id="reactionAs-tooltip" class="help-tooltip" href="#" 
                                              title='Reactions can be either edges connecting substrates with products or nodes connected to them. Reactions as edges makes analysis more flux-centric.'>
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                      <div id="reactionsAsHolder" class="shiny-html-output"></div>
                                      <!--
                                      <select id="reactionsAs">
                                          <option value="edges" selected="selected">As edges</option>
                                          <option value="nodes">As nodes</option>
                                      </select>
                                      -->
                                      <div class="form-group shiny-input-container" data-display-if="input.reactionsAs == 'nodes'">
                                        <div class="checkbox">
                                          <label for="collapseReactions">
                                              <input id="collapseReactions" type="checkbox" checked="checked"/>
                                              <span>Collapse reactions
                                                  <a id="collapseReactions-tooltip" class="help-tooltip" href="#" 
                                                      title='Collapse reaction nodes with the same enzyme.'>
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                          </div>
                                      </div>
                                      <div data-display-if="input.reactionsAs != 'nodes'">
                                          <label class="checkbox" for="useRpairs">
                                              <span>Use RPAIRs
                                                  <a id="useRpairs-tooltip" class="help-tooltip" href="#" 
                                                      title='Connect only those substrate-product pairs that make up a KEGG RPAIR.'>
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                          <input id="useRpairs" type="checkbox" checked="checked"/>
                                      </div>
                                      <button id="preprocess" type="button" class="btn action-button">Make network</button>
                                  </form>
                              </div>
                              <div class="col-md-9">
                                  <div>
                                      <h3>Genomic data</h3>
                                      <div id="geneDESummary" class="shiny-html-output"></div>
                                      <div id="geneDETable" class="shiny-html-output"></div>
                                  </div>
                                  <div>
                                      <h3>Metabolic data</h3>
                                      <div id="metDESummary" class="shiny-html-output"></div>
                                      <div id="metDETable" class="shiny-html-output"></div>
                                  </div>
                                  <div class="bottom-buffer">
                                      <h3>Network summary</h3>
                                      <div id="networkSummary" class="shiny-html-output"></div>
                                      <div data-display-if="network.available">
                                          <a id="downloadNetwork" class="btn shiny-download-link" href="" target="_blank">Download XGMML</a><br>
                                      </div>
                                  </div>
                              </div>
                          </div>
                          <div id="module-panel" class="row top-buffer">
                              <div class="col-md-3">
                                  <form class="well">
                                      <p class="form-header">Discover modules</p>
                                      <div data-display-if="network.hasGenes">
                                          <label for="geneFDR">FDR for reactions
                                              <a id="geneFDR-tooltip" class="help-tooltip" href="#" 
                                                  title='Increasing/decreasing this value makes adding reactions to a module harder/easier.'>
                                                  <span class="glyphicon glyphicon-question-sign"/>
                                              </a>
                                          </label>
                                          <input id="geneFDR" type="number" value="1e-06" min="1e-100" max="1" step="1e-200"/>
                                      </div>

                                      <label for="metFDR">FDR for metabolites
                                          <a id="metFDR-tooltip" class="help-tooltip" href="#" 
                                              title='Increasing/decreasing this value makes adding metabolites to a module harder/easier.'>
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                      <input id="metFDR" type="number" value="1e-06" min="1e-100" max="1" step="1e-200"/>
                                      <!--<input id="metMinusLogFDR" name="metMinusLogFDR" class="minusLogFDR" type="slider" value="5" data-skin="plastic"/>-->

                                      <label for="absentMetScore">Score for absent metabolites
                                          <a id="absentMetScore-tooltip" class="help-tooltip" href="#" 
                                              title='Increasing/decreasing this value makes adding metabolites with no data to a module harder/easier.'>
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                      <input id="absentMetScore" type="number" value="-20"/>

                                      <div data-display-if="!network.hasGenes">
                                          <label for="absentRxnScore">Score for absent reactions
                                              <a id="absentRxnScore-tooltip" class="help-tooltip" href="#" 
                                                  title='Increasing/decreasing this value makes adding reactions with no data to a module harder/easier (usable only when there is no gene expression data).'>
                                                  <span class="glyphicon glyphicon-question-sign"/>
                                              </a>
                                          </label>
                                          <input id="absentRxnScore" type="number" value="-10"/>
                                      </div>

                                      <label class="control-label" for="solver">Select a solver:
                                          <a id="selectSolver-tooltip" class="help-tooltip" href="#" 
                                              title='Multiple MWCS solvers can be used to find modules. Check the help page for details.'>
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>

                                      <select id="solver">
                                          <option value="heinz" selected="selected">Heinz</option>
                                          <option value="mwcs">MWCS (under development)</option>
                                          <option value="fastHeinz">Heuristic search</option>
                                      </select>

                                      <div data-display-if="$('#solver').val() == 'mwcs'">
                                          <label for="mwcsTimeLimit">MWCS solver time limit (seconds):
                                          <a id="selectSolver-tooltip" class="help-tooltip" href="#" 
                                              title="It's an approximate maximum time MWCS solver is allowed to run. After hitting the time limit it will print the best solurion found so far.">
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                          <input id="mwcsTimeLimit" type="number" value="10" min="0" max="120"/>
                                      </div>

                                      <div data-display-if="$('#solver').val() == 'heinz'">
                                          <label for="heinzTimeLimit">Heinz time limit (seconds):
                                          <a id="selectSolver-tooltip" class="help-tooltip" href="#" 
                                              title="It's an approximate maximum time heinz solver is allowed to run. After hitting the time limit it will just stop searching outputing no module.">
                                              <span class="glyphicon glyphicon-question-sign"/>
                                          </a>
                                      </label>
                                          <input id="heinzTimeLimit" type="number" value="60" min="0" max="240"/>
                                      </div>
                                      <button id="find" type="button" class="btn action-button">Find module</button>
                                      <div data-display-if="network.hasReactionsAsEdges && network.usesRpairs">
                                          <label class="checkbox" for="addTransPairs">
                                              <input id="addTransPairs" type="checkbox"/>
                                              <span>Add trans- edges
                                                  <a id="addTransPairs-tooltip" class="help-tooltip" href="#" 
                                                      title='Add trans- RPAIR edges.'>
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                      </div>
                                      <div data-display-if="network.hasReactionsAsNodes">
                                          <label class="checkbox" for="addMetabolitesForReactions">
                                              <input id="addMetabolitesForReactions" type="checkbox"/>
                                              <span>Add all metabolites for reactions
                                                  <a id="addMetabolitesForReactions-tooltip" class="help-tooltip" href="#" 
                                                      title='Add all metabolites that are part of any reaction in the module.'>
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                          <label class="checkbox" for="addInterconnections">
                                              <input id="addInterconnections" type="checkbox"/>
                                              <span>Add interconnections
                                                  <a id="addInterconnections-tooltip" class="help-tooltip" href="#" 
                                                      title="Addg all reactions that aren't present in the module but are directly connected to two metabolites from the module.">
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                          <label class="checkbox" for="removeHangingNodes">
                                              <input id="removeHangingNodes" type="checkbox"/>
                                              <span>Remove hanging nodes
                                                  <a id="removeHangingNodes-tooltip" class="help-tooltip" href="#" 
                                                      title='Remove all metabolites with no data and only one connection in the module.'>
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                          <label class="checkbox" for="simplifyReactionNodes">
                                              <input id="simplifyReactionNodes" type="checkbox"/>
                                              <span>Simplify reaction nodes
                                                  <a id="simplifyReactionNodes-tooltip" class="help-tooltip" href="#" 
                                                      title='Replacing reaction nodes with edges for reactions that connect only two metabolites in the module and these two metabolites are on different sides of the reaction.'>
                                                      <span class="glyphicon glyphicon-question-sign"/>
                                                  </a>
                                              </span>
                                          </label>
                                      </div>
                                  </form>
                                  <h3>Module summary</h3>
                                  <div id="moduleSummary" class="shiny-html-output"></div>
                                  <div data-display-if="module.available">
                                      <a id="downloadModule" class="btn shiny-download-link" href="" target="_blank">Download XGMML</a>
                                  </div>
                              </div>
                              <div class="col-md-9">
                                  <div id="legend">
                                      <div style="color: red">Red: log2FC &gt; 0</div>
                                      <div style="color: green">Green: log2FC &lt; 0</div>
                                      <div style="color: #00acad">Aqua: log2FC not availavle</div>
                                  </div>
                                  <div id="module" class="graph-output"></div>
                              </div>
                          </div>
                      </div>
                      <div class="tab-pane" id="tab-help">
                          <div class="row">
                              <div class="col-md-9 offset3">
                                  <p>This application is a web-interface to the <a href="bioconductor/GAM"
target="_blank">GAM</a> (Gene And Metabolic data analysis) R-package. GAM analyzes gene and/or metabolic differential expression (DE) data in a context of biochemical reactions.</p>
<p>Analysis consists of creating a network describing connections between metabolites and reactions specific to the data and finding a connected module that contains the most significant changes.</p>
<h4 id="constructing-a-network">Constructing a network</h4>
<p>To construct a network select an organism from the <em>Select an organism</em> dropdown menu. Only reactions possible in the selected organism are used. Currently supported organisms are Homo sapiens and Mus musculus. If your organism is not supported, please contact <a
href="mailto:asergushichev@path.wustl.edu">asergushichev@path.wustl.edu</a>.</p>
<!-- :ToDo: add a screenshot? -->

<p>The next step is to upload gene and/or metabolite DE data. GAM can be run using either gene DE data or metabolite DE data or both datasets. Each DE dataset must be in a separate CSV file (comma-, tab- and space- separated files are supported, archived files are supported too). The first line of each file must contain a header with column names. Files should contain the following columns:</p>
<ul>
<li>&quot;ID&quot;: RefSeq mRNA transcript ID, Entrez ID or symbol for genes and and HMDB or KEGG ID for metbolites.</li>
<li>&quot;pval&quot; for the p-value;</li>
<li>&quot;log2FC&quot; for the base 2 logarithm of the fold-change.</li>
</ul>
<p><strong>NB: Columns in your files can have a little different names and GAM will try to guess which one to use.</strong> If you favorite DE tool produce names that GAM can not recoginze, please tell me and I will try to add it.</p>
<p>The &quot;log2FC&quot; column is optional, but we recommend to provide it if possible, as it's used for colors in graph visualizing. Any other columns will be copied to a network as node or edge attributes. Example data for genes and metabolites can be downloaded <a href="/publications/supp_materials/GAM/">here</a>.</p>
<p>After files are uploaded, a file summary is displayed. Verify that the files were parsed correctly.</p>
<div class="figure">
<img src="img/data_summary.png" alt="Summary of the example data" /><p class="caption">Summary of the example data</p>
</div>
<p>If gene DE data was uploaded, the gene DE data is converted to reaction DE data. This is done by considering for a reaction all the genes that code any enzyme that takes part in this reaction. P-values of these genes are compared. Gene with the minimal p-value is selected and its p-value is assigned as the reaction p-value. All reactions without p-values are discarded as having no expressed enzymes. We recommend to exclude genes with low expression prior uploading.</p>
<p>Reactions in the network can be interpreted as either edges or nodes. Select one of the interpretations.</p>
<p><strong>Reactions as edges.</strong> If reactions are interpreted as edges, if a pair of metabolites has multiple reaction connections between them, only the reaction with the minimal p-value is kept. All the remaining reactions and metabolites that take part in these reactions are combined into a network as edges and nodes respectively. To only consider cross-connections between substrates and products in reactions that make up a KEGG reaction pair (RPAIR) as opposed to all cross-connections, select the <em>Use RPAIRs</em> option (seeÂ <a href="http://www.genome.jp/kegg/reaction/">KEGG REACTION</a> for details).</p>
<p><strong>Reactions as nodes.</strong> If reactions are interpreted as nodes, then both metabolites and reactions are added as nodes to the network. Edges are added between a metabolite and a reaction if the metabolite takes part in this reaction. To collapse groups of reactions that have at least one common metabolite and the same most significant gene into single nodes, select the <em>Collapse reactions</em> option.</p>
<p>Generally, all of these options will lead to similar results. We recommend to use the default option values as this makes the network simpler and the analysis faster.</p>
<h4 id="finding-a-module">Finding a module</h4>
<p>After making a network, you can find a connected module that contains the most significantly changed genes and reactions. Internally, this is done by first scoring nodes and edges based on their p-values such that positive scores correspond to significant p-values and negative scores correspond to insignificant changes. Then the problem of finding a connected subgraph with maximum summary weight (maximum-weight connected subgraph, MWCS problem) is solved.</p>
<p>This site supports three solvers:</p>
<ul>
<li>The <a href="http://www.mi.fu-berlin.de/w/LiSA/Heinz">Heinz solver</a>: it's the default solver, as it works for both interpretations of reactions.</li>
<li>The MWCS solver: this solver can provide more interactive experience because when reaching time limit it outputs the best solution found so far.</li>
<li>The heuristic search: this solver is from R-package BioNet. Added for comparison, as it's distributed freely.</li>
</ul>
<p>As the last two solvers don't support edge scores they are not available when reactions are interpreted as edges.</p>
<p>The <em>FDR values</em>, <em>Score for absent metabolites</em>, and <em>Score for absent reactions</em> options control the size of the module. Increasing/decreasing the <em>FDR for reactions</em> (<em>FDR for metabolites</em>) value makes adding reactions (metabolites) to a module harder/easier. We recommend to start from the default values and then gradually change them depending on the results.</p>
<p>Click &quot;Find module&quot; button to find a module in the network. The module will be shown on the right panel.</p>
<div class="figure">
<img src="img/module.png" alt="Example of a module" /><p class="caption">Example of a module</p>
</div>
<h4 id="graph-legend">Graph legend</h4>
<p>We use the following scheme:</p>
<ul>
<li>Red nodes and edges are up-regulated.</li>
<li>Green nodes and edges are down-regulated.</li>
<li>Blue nodes and edges don't have <em>log2FC</em> values.</li>
<li>Bigger size of nodes and width of edges means lower p-values.</li>
<li>Dashed edges are trans-RPAIRs.</li>
</ul>
<h4 id="post-processing">Post-processing</h4>
<p>There are couple of post-processing steps that are available. If in the network reactions are edges and RPAIRs are used, then trans-connections can be added. When reactions are nodes, the following operations can be applied:</p>
<ul>
<li>Adding all metabolites that are part of any reaction in the module.</li>
<li>Adding all reactions that aren't present in the module but are directly connected to two metabolites from the module.</li>
<li>Removing all metabolites with no data and only one connection in the module.</li>
<li>Replacing reaction nodes with edges for reactions that connect only two metabolites in the module and these two metabolites are on different sides of the reaction.</li>
</ul>
<h4 id="saving-the-module">Saving the module</h4>
<p>You can download the module in an XGMML format by clicking the <em>Download XGMML</em> button. This file can be imported into Cytoscape. You can also download GAM's VizMap style for Cytoscape <a href="/publications/supp_materials/GAM/GAM_VizMap.xml">here</a>.</p>

                              </div>
                          </div>
                      </div>
                      <div class="tab-pane" id="tab-about">
                          <div class="row">
                              <div class="col-md-9 offset3">
                                  <p> By Alexey Sergushichev (<a href="mailto:asergushichev@path.wustl.edu">asergushichev@path.wustl.edu</a>)</p>
                                  <p> This site must not be used for commercial purposes.<p>
                                  <p> You can find source code of this site <a href="https://bitbucket.org/assaron/gam">here</a>.</p>
                                  <div id="GAMVersion" class="shiny-html-output"></div>
                                  <p> Links <p>
                                  <p> Powered by Shiny. </p>
                              </div>
                          </div>
                      </div>
                  </div>
              </div>
          </div>
      </div>
    </div>
    <div id="showModulePanel" class="js-output"></div>
  </body>
</html>

